### ISSUE TITLE
Inventory system

### ISSUE THREAD

**Issue by spencer-melnick** [spencer-melnick/Threshold#22]
Create a simple inventory system. Inventory can be all server-side authoritative with little to no prediction.

Inventory items should be able to apply modifiers like defense and attack bonuses to the character while they are equipped.

**Comment by spencer-melnick** [comment]
It might be beneficial to move the ability system component of the character into the player state at this point in time, since the inventory will also be stored in the player state.

I'm thinking that all inventory objects should inherit from an interface, that way generic inventory items can be derived from UDataAsset, but more complex inventory items could inherit from a different class type as needed.

**Comment by spencer-melnick** [comment]
I put in a little TODO in the inventory code (I still need to test it), that the client should disconnect if the server starts replicating garbage data to the inventory (invalid item classes mainly).

The steps for disconnecting require cleaning up network resources (sessions, not sure what else), and then deleting the PlayerController and Pawn. The error message could probably be stored in some GameInstance state, and the forced disconnect method could be added to a subsystem. Might be a good idea to add a separate issue? TBD

**Comment by spencer-melnick** [comment]
I changed the system to replicate object references so it would be possible to create objects using a data table. Currently, the inventory subsystem scans the directory specified in DefaultGame.ini for data tables to build the inventory objects.

However, currently the object references are not replicatable across the network, so I will have to investigate a little further.

**Comment by spencer-melnick** [comment]
Another major change to the system - now inventory items are held in a simple struct, supporting polymorphic replication via a handle (much how GAS handles target data).

This allows replicating very minimal amounts of data for simple inventory items created via a table (replicating the source table and row name hash only), or replicating all of their data as needed. I've included some example item types to demonstrate how to set them up (there is a lot of boilerplate!).

I managed to allow assigning any inventory item type in editor via a very simple hack, by using an instanced UObject that only has one function `FInventoryItem* GetItem()`.

Lastly I just need to add accessors with stacking support, removal support, etc.
